src/chkstat.cpp

```diff
@@ -21,1365 +21,1246 @@
+
+            link.resize(PATH_MAX);
+            const auto len = ::readlinkat(pathfd.get(), "", &link[0], link.size() - 1);
+
+            if (len <= 0 || static_cast<size_t>(len) >= link.size() - 1)
+                // TODO: would an error read the link not warrant a complaint?
+                return false;
+
+            link.resize(static_cast<size_t>(len));
+            stripTrailingSlashes(link);
+
+            if (link[0] == '/')
             {
-                BAD_LINE();
-                continue;
+                // absolute link, need to continue from a new root
+                pathfd.close();
             }
-            *part[2]++ = 0;
-            mode = (mode_t)strtoul(part[3], part + 3, 8);
-            if (mode > 07777 || part[3][0])
+            else
             {
-                BAD_LINE();
-                continue;
+                // relative link: continue relative to the parent directory
+
+                // we encountered a link directly below /, simply continue
+                // from the root again
+                if (parentfd.invalid())
+                {
+                    pathfd.close();
+                }
+                else
+                {
+                    pathfd.set( dup(parentfd.get()) );
+                }
             }
-            last = add_permlist(part[0], part[1], part[2], mode);
+
+            path_rest = link + path_rest;
+        }
+        else if (ctx.status.isDirectory())
+        {
+            // parentfd is only needed to find the parent of a symlink.
+            // We can't encounter links when resolving '.' or '..' so those don't need any special handling.
+            parentfd.set( dup(pathfd.get()) );
         }
-        fclose(fp);
     }
 
-    euid = geteuid();
-    for (e = permlist; e; e = e->next)
+    // world-writable file: error out due to unknown file integrity
+    if (ctx.status.isRegular() && ctx.status.isWorldWritable())
     {
-        if (use_checklist && !in_checklist(e->file+rootl))
-            continue;
+        std::cerr << ctx.subpath << ": file has insecure permissions (world-writable)" << std::endl;
+        return false;
+    }
 
-        pwd = strcmp(e->owner, "unknown") ? getpwnam(e->owner) : NULL;
-        grp = strcmp(e->group, "unknown") ? getgrnam(e->group) : NULL;
-        uid = pwd ? pwd->pw_uid : 0;
-        gid = grp ? grp->gr_gid : 0;
+    ctx.fd.steal(pathfd);
 
-        bool traversed_insecure;
-        if (fd >= 0)
-        {
-            // close fd from previous loop iteration
-            close(fd);
-            fd = -1;
-        }
+    return true;
+}
+
+std::string Chkstat::getPathFromProc(const FileDesc &fd) const
+{
+    std::string linkpath;
+    auto procpath = std::string("/proc/self/fd/") + std::to_string(fd.get());
+
+    linkpath.resize(PATH_MAX);
 
-        fd = safe_open(e->file, &stb, uid, &traversed_insecure);
-        if (fd < 0)
+    ssize_t l = readlink(procpath.c_str(), &linkpath[0], linkpath.size() - 1);
```

Comment:
C++ doesn't require a NUL byte in the end, so no need to subtract 1 from the string length (also two lines below).